### PR TITLE
fix(input): support alt-left-bracket

### PR DESCRIPTION
**Pull request by imsnif**
Hey,

So the story is I'd really like to add the `Alt` + `[` keybind to Zellij, but right now termwiz breaks it down to `Esc` and `[`, even when `maybe_more` is false.

This fixes it, but I'm pretty sure I did the wrong thing here - not only because the solution is rather naive but also because some relevant tests are failing.

If you could point me in the right direction to make this right, that would be greatly appreciated.

Thanks!

### CODE REVIEW

**Comment by wez** [comment]
Could you add a test case for the specific issue you're fixing?
I think something similar to the `partial_ambig` might be roughly the right thing.
That will help me form a clearer picture on what we're shooting for and then we can figure out how to get everything playing nicely together!

**Comment by wez** [comment]
(and thanks for diving in!)

**Comment by imsnif** [comment]
img-2023-01-25-173634

I see the CI does not take any prisoners ;P

> (and thanks for diving in!)

Gladly! I just added a test - what do you think?

**Comment by wez** [comment]
OK, I see what the issue is, and the fix isn't in the right place.

First, I think that we need to add an explicit entry into the key map for `Alt-[`:

```diff
--- a/termwiz/src/input.rs
+++ b/termwiz/src/input.rs
@@ -1453,6 +1453,13 @@ impl InputParser {
                 modifiers: Modifiers::NONE,
             }),
         );
+        map.insert(
+            b"\x1b[",
+            InputEvent::Key(KeyEvent {
+                key: KeyCode::Char('['),
+                modifiers: Modifiers::ALT,
+            }),
+        );

         map
     }
```

that teaches the map that that is a valid key combination, but we get failures in the tests because it's an especially problematic mapping, as it matches the prefix of very many other escape sequences.

What I think the fix needs to be is along the lines of:

* Pass the `maybe_more` parameter to `Node::lookup` in `termwiz/src/keymap.rs` so that the lookup logic in there can reason about the ambiguity
* In the case where the lookup would return `AmbiguousMatch`, if `maybe_more == false`:
   * Look for the longest prefix of `key` that matches a single node. If found, return that as an exact match.
   * Otherwise it is ambiguous and we should return the existing ambiguous match result

Currently it returns the shortest match and the `partial_ambig` test case fails because it ends up seeing `Alt-[` instead of `ESC[11`; we need the longest match for that test to pass.

**Comment by imsnif** [comment]
Thanks for the rundown!

I played around with it a little and I think I understand what you're getting at. But what I'm still a little confused about is:

The test checks for: `[27, 91, 49, 49]` (`ESC`, `[`, `1`, `1`)
It expects to get all of these as separate events*, which is the case currently. With the above solution, we would still get [`Alt-[`, `1`, `1`] because `[27, 91]` is longer than `[27]`. 

What am I missing?

*The comments mention it's expecting an F-key, but the assertion seems to look for separate events

**Comment by imsnif** [comment]
If I'd have to guess at a solution (without knowing the particularities of the parser, ofc) I'd go for something like:

Identify this particular key combination as a separate entity, only matching against it if `maybe_more` is false and the buffer contains nothing else. Which I think will always be the case with this sort of user input (unless there's some external automation involved).

**Comment by wez** [comment]
> The comments mention it's expecting an F-key, but the assertion seems to look for separate events

Looking more closely, the comment `// Fragment this F-key sequence across two different pushes` is copied from the `partial` test above it, but it doesn't call out the difference between the two tests, which is that the `partial` finishes off the CSI-based F1 key sequence with the trailing `~`, whereas `partial_ambig` is correct in producing the individual input events because the sequence that gets finished by the second `p.parse(..., false)` call doesn't match any valid key in the keymap.

I think that misleading comment should be edited to call out that it is an incomplete F-key sequence.

For a bit of background, the keymap datastructure is a relatively simple prefix tree; using the `lookup` test case as an illustrative example, it inserts `boa` -> `true`, `boo` -> `true` and `boom` -> `false` into the tree. The resultant tree would look something like:

```mermaid
flowchart TD
    ROOT[keymap] --> B1[b]
    B1 --> O1[o]
    O1 --> O2[o]
    O2 -- value --> O2V[true]
    O2 --> M1[m]
    M1 -- value --> M1V[false]
    O1 --> A1[a]
    A1 -- value --> A1V[true]
```

> Identify this particular key combination as a separate entity, 

Yep; that `map.insert` I suggested above would do that bit

> only matching against it if `maybe_more` is false and the buffer contains nothing else.

Yeah, but note that the buffer might have more than just that one key press contained within it though, if the user is typing quickly, so I think a more robust check would be to find the longest prefix of `buf` that resolves to an unambiguous entry in the keymap.

Right now the lookup routines for the keymap assume that we will always want the shortest matching prefix.

I think that maybe adding a separate `lookup_longest` method might be a bit easier to compose with the rest of the code to achieve the desired outcome; maybe the `lookup` method doesn't need to take to the `maybe_more` parameter directly, and the decision to use `lookup_longest` could bubble up a bit higher, if that helps make the control flow easier to follow.

**Comment by imsnif** [comment]
Thanks for the detailed explanation (and the nice looking mermaid diagram!) - I think I understood this before, but I'll just rephrase it to make sure we're on the same page:

You're suggesting that given a complete buffer (i.e. `maybe_more` is `false`), we adjust the lookup function to look for the longest match instead of the shortest as we do now. So that `[<ESC>, '[']` would yield `ALT-[` instead of `ESC` + `[`. Right?

This much I implemented the other day and it seems to work great. However, there's another issue I'm either being dense about or not managing to explain properly. So let me try again:

With this implementation, let's look at the `partial_ambig` test. This test sends the input: `<ESC>[11` and is expecting to get the events:
    `Key(KeyEvent { key: Escape, modifiers: NONE })`,
    `Key(KeyEvent { key: Char('['), modifiers: NONE })`, 
    `Key(KeyEvent { key: Char('1'), modifiers: NONE })`,
    `Key(KeyEvent { key: Char('1'), modifiers: NONE })`


With the proposed solution we're talking about, what we get for this test is instead:
`Key(KeyEvent { key: Char('['), modifiers: ALT })`, 
`Key(KeyEvent { key: Char('1'), modifiers: NONE })`,
`Key(KeyEvent { key: Char('1'), modifiers: NONE })]`

`<ESC>[` could be either `Alt-[` or `ESC` + `[`, since the first one is a longer match, we choose it. But this is almost certainly not what we should be doing from a user perspective (I think!).

I can go into more details about my proposed solution to this, but before I do: does this make sense? Or am I missing something? Maybe there's some detail I still did not understand.

**Comment by imsnif** [comment]
Actually, thinking further about this I realize I'm very much looking at this perspective from a "reading from local STDIN" perspective. Which colors my vision a bit and makes me ignore the heavily fragmented realities of eg. SSH connections.

Meaning the ambiguity of `ESC` + `[` vs. `Alt-[` is something we have to live with and can't do a whole lot about.

So if you're good with me implementing your "longest match" solution and changing the above test to expect the `Alt-[` event rather than the two events (`ESC` and `[`), I have no further qualms. What say you?

**Comment by wez** [comment]
> You're suggesting that given a complete buffer (i.e. `maybe_more` is `false`), we adjust the lookup function to look for the longest match instead of the shortest as we do now. So that `[<ESC>, '[']` would yield `ALT-[` instead of `ESC` + `[`. Right?

The precise logic I had in mind was: if `maybe_more` is `false`, if the regular lookup is ambiguous, then try a lookup using the longest matching prefix from the buffer. If that returns an unambiguous result, then that is our result.

However, I suppose that initial shortest prefix lookup would be totally redundant with the longest prefix lookup in that case, so we may as well skip it.

> With the proposed solution we're talking about, what we get for this test is instead:
> `Key(KeyEvent { key: Char('['), modifiers: ALT })`,
> `Key(KeyEvent { key: Char('1'), modifiers: NONE })`,
> `Key(KeyEvent { key: Char('1'), modifiers: NONE })]`

Yeah, this popped into my head just before I went to bed. I think this result with it yielding `Alt-[` is correct with the changes we're proposing, because the full buffer contents in that test case cannot ever match anything longer (as `maybe_more==false` on the second parse call).

So, let's update the expectations for the `partial_ambig` test!

Thanks again for looking at this!

**Comment by imsnif** [comment]
Just to keep you and this thread in the loop: I think I'm good to proceed with the implementation. Will hopefully get to this ASAP, but not 100% sure when.

Either way: this is part of the requirements for the next Zellij release which I'm spending all my time on. So this will certainly be happening and not forgotten!

**Comment by imsnif** [comment]
Hey @wez - so I finally got around to this.

I made all the changes we talked about, namely:
1. We now pass `maybe_more` to the `lookup` function (including updating all the relevant tests)
2. When `maybe_more` is `false`, the `lookup` function will return the longest `AmbiguousMatch` as an `ExactMatch`
3. I added two additional tests to `KeyMap` to specify the old behaviour with `maybe_more` true and the new behaviour with `maybe_more` false.
4. I adjusted the comment in the `partial_ambig` tests as well as its expected results

Please let me know if you'd like anything changed, and I hope you have a nice weekend :)

File: termwiz/src/keymap.rs [hunk 1]
```diff
@@ -60,12 +66,18 @@ impl<Value: Debug> Node<Value> {
...
                     NodeFind::AmbiguousBackTrack => {
                         // The child didn't have an exact match, so check
                         // whether we do
                         match self.value.as_ref() {
-                            Some(value) => NodeFind::AmbiguousMatch(depth, value),
+                            Some(value) => {
+                                if maybe_more {
+                                    NodeFind::AmbiguousMatch(depth, value)
+                                } else {
+                                    NodeFind::Exact(depth, value)
```
[the remark this answers is not in the document]
Comment (imsnif) [reply termwiz/src/keymap.rs:83]: Sure, I added a comment here. I hope it's clear enough?

**Comment by imsnif** [comment]
Any hint about the tests? I can't seem to find the files they are pointing at and locally everything passes.

**Comment by wez** [comment]
Looks like it might have tried to merge your stuff onto main; if you rebase your branch onto the current wezterm main you'll see that line 2070 needs updating to pass in the maybe_more parameter.

**Comment by imsnif** [comment]
> Looks like it might have tried to merge your stuff onto main; if you rebase your branch onto the current wezterm main you'll see that line 2070 needs updating to pass in the maybe_more parameter.

Got it, thanks!

I fixed this but unfortunately made a git salad by accident. Let me know if you'd like me to just trash this and do a new PR.

**Comment by wez** [comment]
Ah, yeah, I think you merged the wrong way around :-(
I can't merge this PR as-is because my repo policy is to only accept fast-forwardable merges to keep the history linear.
You can either fix this up and force-push, or just make a new PR if that is easier to wrangle: git can be difficult to master!

**Comment by imsnif** [comment]
Fixed! I ended up force-pushing as you suggested :)

**Comment by wez** [comment]
Thank you!

**Comment by wez** [comment]
Published in termwiz 0.20

**Comment by imsnif** [comment]
Thank you as well for the super quick turnaround!
